### PR TITLE
feat: implement conflict checker

### PR DESCRIPTION
**Pull request by bsbds**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

  Implements the conflict checker. It is a replacement for the `conflict_checked_mpmc in the upcoming refactor.

* what changes does this pull request make?

  Implements the conflict checker

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

  No

### CODE REVIEW

File: curp/src/server/conflict_checker.rs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+                    return true;
+                };
+                let EntryData::Command(ref cmd2) = entry2.entry_data else {
+                    return true;
+                };
+                cmd1.is_conflict(cmd2)
+            }
+            _ => true,
+        }
+    }
```
Comment (themanforfree) [inline curp/src/server/conflict_checker.rs:170]: Please use `ConflictCheck` trait

File: curp/src/server/conflict_checker.rs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+    /// Registers a vertex to the checker
+    pub(crate) fn register(&self, vertex: VertexType<C>) -> Registration<C, Waiting> {
+        let new_vid = self.next_vertex_id();
+        match vertex {
+            VertexType::Entry(ref e) => {
+                // Already registered in speculative execution
+                if let Some(entry) = self.cmd_vid.get(&e.propose_id) {
+                    return self.new_register(*entry.value(), None);
+                }
```
Comment (themanforfree) [inline curp/src/server/conflict_checker.rs:91]: if vertex is already registered, we needn't generate a new_vid in line 85

File: curp/src/server/conflict_checker.rs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    /// Conflict graph
+    vs: Arc<DashMap<u64, Vertex<C>>>,
+    /// Next vertex id
+    next_id: AtomicU64,
+}
+
+/// The vertex type stored in conflict checker
+#[derive(Debug)]
+pub(crate) enum VertexType<C: Command> {
+    /// A entry vertex
```
Comment (Phoenix500526) [inline curp/src/server/conflict_checker.rs:34]: It's "An", not "A"

File: curp/src/server/conflict_checker.rs [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+    #[tokio::test]
+    async fn conflict_checking_for_pairs_should_work() {
+        let mut gen = EntryGen::default();
+        let entry_pairs_conflict = [
+            (gen.log_entry(vec![0]), gen.log_entry(vec![0])),
+            (gen.log_entry(vec![0]), VertexType::Snapshot),
+            (VertexType::Snapshot, VertexType::Snapshot),
+            (VertexType::Reset, gen.log_entry(vec![0])),
+            (VertexType::Reset, VertexType::Snapshot),
```
Comment (Phoenix500526) [inline curp/src/server/conflict_checker.rs:234]: Is here missing the `(VertexType::Reset, VertexType::Reset)` case?

**Comment by bsbds** [comment]
The conflict checker seems no longer needed in the upcoming refactor, I'll convert it to draft for now.

**Comment by bsbds** [comment]
This PR is no longer required due to another refactor.

**Pull request closed without merging**